spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskExecutionService.java

```diff
@@ -179,64 +190,77 @@ public long executeTask(String taskName, Map<String, String> taskDeploymentPrope
+		AppDeploymentRequest appDeploymentRequest = this.taskAppDeploymentRequestCreator.
+				createRequest(taskExecution, taskExecutionInformation, commandLineArgs, platformName);
 
-			final String dataflowServerAccessTokenKey = "dataflow-server-access-token";
-			final String dataflowServerUseUserAccessToken = "dataflow-server-use-user-access-token";
+		TaskManifest taskManifest = createTaskManifest(platformName, taskExecutionInformation, appDeploymentRequest);
 
-			for (String commandLineArg : commandLineArgs) {
-				if (commandLineArg.startsWith("--" + dataflowServerAccessTokenKey)) {
-					containsAccessToken = true;
-				}
-				if (StringUtils.trimAllWhitespace(commandLineArg).equalsIgnoreCase("--" + dataflowServerUseUserAccessToken + "=true")) {
-					useUserAccessToken = true;
-				}
-			}
+		TaskManifest previousManifest = this.dataflowTaskExecutionMetadataDao.getLastManifest(taskName);
 
-			final String dataflowAccessTokenPropertyKey = "app." + taskExecutionInformation.getTaskDefinition().getRegisteredAppName() + "." + dataflowServerAccessTokenKey;
-			for (Map.Entry<String, String> taskDeploymentProperty : taskExecutionInformation.getTaskDeploymentProperties().entrySet()) {
-				if (taskDeploymentProperty.getKey().equals(dataflowAccessTokenPropertyKey)) {
-					containsAccessToken = true;
-				}
+		if(!taskDeploymentProperties.isEmpty()) {
+			taskDeploymentProperties = appDeploymentRequest.getDeploymentProperties();
+		}
+		else {
+			if(previousManifest != null && !previousManifest.getTaskDeploymentRequest().getDeploymentProperties().equals(taskDeploymentProperties)) {
+				appDeploymentRequest = updateDeploymentProperties(commandLineArgs, platformName, taskExecutionInformation, taskExecution, previousManifest);
+
+				taskDeploymentProperties = previousManifest.getTaskDeploymentRequest().getDeploymentProperties();
 			}
+		}
 
-			if (!containsAccessToken && useUserAccessToken) {
-				final String token = TokenUtils.getAccessToken();
+		AppDeploymentRequest request = new AppDeploymentRequest(appDeploymentRequest.getDefinition(),
+				appDeploymentRequest.getResource(),
+				taskDeploymentProperties,
+				appDeploymentRequest.getCommandlineArguments());
+
+		taskManifest.setTaskDeploymentRequest(request);
+
+		if(!isAppDeploymentSame(previousManifest, taskManifest)) {
+
+			validateAndLockUpgrade(taskName, platformName, taskExecution);
+
+			logger.debug("Deleting %s and all related resources from the platform", taskName);
+			taskLauncher.destroy(taskName);
 
-				if (token != null) {
-					taskExecutionInformation.getTaskDeploymentProperties().put(dataflowAccessTokenPropertyKey, token);
-				}
-			}
 		}
 
-		AppDeploymentRequest request = this.taskAppDeploymentRequestCreator.
-				createRequest(taskExecution, taskExecutionInformation, commandLineArgs, platformName);
+		this.dataflowTaskExecutionMetadataDao.save(taskExecution, taskManifest);
+
+		String taskDeploymentId = taskLauncher.launch(appDeploymentRequest);
```

Comment:
Since we already destroyed the task above, what happens to the upgrade(and the state we maintain in the following steps) when there is a failure at this step?

Reply:
I think we need to handle possible exceptions and maintain the state of the upgrade accordingly.

Reply:
I'll address this a bit as mentioned in my comment to @jvalkeal 's comment.  However, do keep in mind that the `TaskLauncher` will deploy a new instance if one is not there so the "recovery" should be to just try again in most scenarios.  The only thing I believe I'm missing (and I'll address as I address the code review comments) is the "rollback" of the in memory lock.